### ISSUE TITLE
Notification of payment in flight

### ISSUE THREAD

**Issue by Pongch** [omisego/elixir-omg#565]
**Story:**
As a merchant, I want to know that I should expect a payment. So I can decide if I want to accept a payment that has been sent to Child chain operator but not yet been included in an Ethereum block
 
**Description**
An event driven mechanism (such as but not restricted to a websocket) in which the beneficiary of a payment transaction is informed that they can expect a payment.

The beneficiary based upon these risk factors can decide if they wish to accept the payment:

- The size of the payment
- The trustworthiness and reliability of the operator

**Background**

When a payment is performed on the OMG Network using js-starter-kit:
`getbalance()` for the beneficiary address does not update until the payment transaction is included in an Ethereum block

There is no visibility of the payment until the payment transaction is included in an Ethereum block. Hence, a notification that the payment is currently in flight allows for the client to deal with Balance UX accordingly (either shows the notification that transaction is made, or optimistically update the balance)

**Comment by Pongch** [comment]
Per discussions with @pdobacz and @pnowosie. On a higher level, there are 3 possible approaches to tackle this story:

1.  Watcher not handling anything not included in the chain, Watcher still maintains it's responsibility of focusing on chain security. Hence the responsibility of tracking of tx happens a service later

2. BYOT (Bring Your Own Transport), similar to MoreVP Tracking of in-flight transaction, albeit with a different focus (UX as Opposed to Security). Possibility of using direct communication protocol like Whisper was mentioned.

3. Transport satisfied by Watcher through W2W (Watcher to Watcher) Communication.

Note: discussion also includes having the Child chain sign an `attestation` that it has received `tx bytes` ..keeping in mind that this does not say anything about the validity of the transaction

**Comment by pdobacz** [comment]
Adding my take on the +'s and -'s, as discussed on call:

1/ 
(+) cleanest responsibilities of the Watcher; 
(-) need to re-implement handling of such txs multiple times, for each application (eWallet, block explorer, direct UI to the Watcher)
2/
(-) Watcher handles txs absent in the chain; 
(+) single, uniform API to ask for data about included/in-flight payments, all applications can call it; 
(+) `transaction.track` endpoint fits in with the MoreVP use case; 
(+) naturally extendable to (3/) if necessary
3/ 
(-) adding W2W communication layer (or worse yet - routing this responsibility through the child chain); 
(-) mixing of concerns and baking in the control and transport of in-flight txs into the Watcher

**Comment by pdobacz** [comment]
Adding some more notes and progress so far. I managed to jot down what I think would be a breakdown of effort involved for the 3 options:

1. in application layer
    a. provide signed attestation of transaction acceptance from `transaction.submit` on ch-ch. Attestation is probably sth like `[blknum, txindex, oindex, raw_txhash]->RLPencode->sign(attestation_priv_key)`. Problem: this suddenly makes the ch-ch require having access to private key, or signing in some other way.
    b. (optional) provide a way to separate the private key to sign attestations from `authority` key that submits child chain blocks
    c. (optional) provide an `transaction.verify_attestation` to verify the attestation in a UX friendly form (uses the child chain authority address to perform the verification!)
2. BYOT
    a. signed attestation from ch-ch (as in 1/)
    b. `transaction.track` endpoint, designated for tracking for UX purposes
    c. question: should `transaction.submit` on the W do `.track` automatically?
    d. question: should `transaction.track` do `verify_attestation` automatically, or do we want the latter endpoint as in (1/). Keep in mind that `transaction.track` for IFE-purposes won't have the attestations, so some optionallity is involved here
    e. extend WatcherDB (postgres) by a flag on transaction table: `:included | :tracked`
    f. implement Postgres migration (all txs seen so far are `:included`)
    g. design/implement presenting of `:tracked` txs in `transaction.get`
    h. design/implement presenting of effect of `:tracked` txs in `account.get_utxos`, `account.get_balance`
    i. think about how this relates to `transaction.create` (I think it shouldn't affect at all)
    j. solve the process of turning a `:tracked` tx into an `:included` one
    k. add a new event emitted in `transfer` channel (so a tracked tx emits two: first `status: :tracked` then `status: :included`, the latter is what we have now)
    l. check other possible implications of holding non-included txs in WatcherDB
    m. solve/discuss problem: tx bytes don't hold details on who spent what. If we need that in the event, we should provide additional info in `transaction.submit` result. The required contents of the notification must be confronted with [the current `transfer` channel events](https://github.com/omisego/elixir-omg/blob/master/docs/api_specs/watcher_websocket_specs.md#events-address_received-and-address_spent)
3. watcher to watcher
    a. all from (3/), except the `transaction.track` endpoint (for now; it's still necessary for IFE tx tracking)
    b. W2W communication, naturally



In 2/ (and 3/), moving away from the design, where the Watcher only handles included transactions, is coming with some increase of complexity - all the small issues to handle 2.b through 2.m.

The question is are we willing to embrace the complexity involved to have this feature.

**Comment by whoisjeremylam** [comment]
Can we consider a new operator run `notifier` service that is considered another layer above - `app level` for the purposes of this discussion?

It's optional for the beneficiary of funds to use this service. The rough flow looks something like this:

1. Beneficiary contacts `notification service` and registers addresses that they are interested in receiving notifications that funds are in flight
1. Sender of funds calls watcher's `submit()` with optional parameter `notify = true`
1. If child chain does not support notification, return some kind of error
1. The child chain contacts the `notification service` and (optionally passes a signed attestation)
1. The `notification service` relays the signed attestation to any subscribed parties

Note the optionality of this notification service. The beneficiary must handle where zero notificatons are available. ie the beneficiary must rely on their watcher. Other non-security related events can be added in the future.

A first version of the this feature probably doesn't need an attestation. This can be added in later if we find use for this feature.

**Comment by boolafish** [comment]
>2. Sender of funds calls watcher's submit() with optional parameter notify = true
>3. If child chain does not support notification, return some kind of error

Personal opinion, if we go with this solution, we can start without consider this. I believe it's fair to just say operator will support notification. 

And a side note, I have a gut feeling the whole notification service can be serverless and be quite cheap.

**Comment by pdobacz** [comment]
OK, so the `notification service`. As I understand this, it is distinct from any other, since it doesn't require attestation (aka promise) from the child chain - all thanks to the fact that the child chain server is the privileged actor who can only push the notifications to the service. Then we still have the part:

> So I can decide if I want to accept a payment that has been sent to Child chain operator

but satisfied differently.

so I see two options then, on top of the first 1.-3. we talked about:
4. Child chain pushes to `notification service`, but it is the application layer that subscribes and handles the notifications (it mirrors (1.) a bit)
5. Child chain pushes to `notification service`, but it is the Watcher that subscribes and handles those notifications (mirrors (2.) a bit). This includes the majority of pain of BYOT aka (2.)

Sounds like by doing 4./5. instead of 1./2. we're prioritizing solving the transport over separation of concerns. Argh, too many options to choose from :thinking:

**Comment by whoisjeremylam** [comment]
To reduce impact to existing code, I feel we could push the event handling through to the application layer. I'd favour this variant:

> 4. Child chain pushes to notification service, but it is the application layer that subscribes and handles the notifications (it mirrors (1.) a bit)

**Comment by Pongch** [comment]
sorry for getting back rather late, but I concur with @whoisjeremylam on 4/ pushing the event handling over to the application, who can subscribe to whichever topic. Also if we pick a more standardized pub/sub then any type of client/app could listen for events

In this case after calling the `.submit()`, the app could subscribe to `.watch('txhash')` to see the state of the current transaction or `.watch('recipient_address')` to see if recipient should expect a transaction.

**Comment by pdobacz** [comment]
Since the exact form of the feature/architecture is still pending, I decided to hack a PoC together, to have _something_ kicked off, should the end of the iteration come.

The simplest that was available to pull off, was to just (incorrectly architecture and responsibilities-wise) allow to connect to a websocket on the child chain server and have that as a medium to transport the notifications.

This is still put together quite hackily, there are some cleanups pending, but the implementation can be looked up here:

https://github.com/omisego/elixir-omg/tree/565-draft_tx_notifications_ws

From here we can do a couple of things:
1. at the very minimum split the application that runs the websocket to a different Phoenix app under the umbrella and decouple. We would then at least have the option to put the notification app on a separate BEAM node (not sure if this is good practice etc)
2. split it further - let the child chain server and such new Phoenix app communicate via the network explicitly (HTTP? WS?)
3. let it talk to some other transport layer (serverless? etc.). Anything you would find a more standard pub sub. (for now the PoC pubsub is a mirror of the Watcher's channels that allow you to get notified about included transactions - the usage and topics and events are exactly the same)

We can also scrap the PoC entirely ofc :). This is just something to get started.

**Comment by pdobacz** [comment]
here is an-`iex` dump of what the notification looks like: 

![Screenshot from 2019-04-19 16-27-36](https://user-images.githubusercontent.com/5735525/56432502-ef0dfd80-62cd-11e9-9a35-b8844cb4057f.png)

**Comment by Pongch** [comment]
hey @pdobacz this is a great PoC !

RE: decoupling of events/notification services, from what you describe 3/ might be ideal as that allows us to extend more convenience functions without putting everything on the same Phoenix app/beam. 

Happy to help draw out the architecture for it this coming week.

**Comment by Pongch** [comment]
Per discussion with @pdobacz today, we came up with a variant of this story that is an extension of 4/ which includes `notification service`. We agreed that this strikes quite a good balance, but any feedbacks are welcome.

**User flow of option 4+** 
1.1 Application subscribes to a transaction made with an address using a watcher
1.2 Watcher subscribes to notification service
2. Application submits a transaction to the watcher as it would normally
3. Notification service emits an event that Childchain has received a transaction
4. Watcher receives that event and relays it to the user.

Pros:
- Watcher not handling anything not included in the chain
- Application still deals with a single endpoint: the watcher 

Cons:
- Application would need to deal with the state of the transaction, transaction is not tracked
- Application would need additional logic to deal with a situation where the transaction is in-flight but not finalized

Let me know if this is not accurate or missing any details.

**Comment by whoisjeremylam** [comment]
Option 4+ looks great @pdobacz and @Pongch!

**Comment by pdobacz** [comment]
Just some smaller comments for full clarity on 4+:
  - the subscriptions done by both the app to the watcher and the watcher to the `notification service` would be done **per address**, that is a `transfer:0x1234...` channel would push notifications about transactions received/sent by `0x1234...`.
  - watcher would (at channel `transfer:0x1234`) push both flavors of notifications, differentiated by a flag (`:pending | :included` for example) to the subscriber. At least this is the plan. As a consequence, the subscription of watcher to `notification_service` would take place on subscription of app to the watcher, hopefully not requiring any additional steps from the app
  - 4+ is naturally extensible to 5, which would add sth called "consolidated API", where `:pending` txs would be possible to be seen via endpoints like `transaction.get/all`, `account.get_utxos`, `account.balance` - similar to BYOT

**Comment by kasima** [comment]
Spoke to @Pongch about the final solution here. From an engineering perspective, it seems like the best balance. This PoC has been a good exploration.

However, we're just beginning to consider work for the next Engineering Cycle – Carl's Jr. 🍔. With it, we want to take another look at the (deprecated) websocket API as a whole, since we've gotten feedback that it's been useful in the Ari testing phase. We're talking about including this feature into the next cycle along with a larger API theme, rather than doing a one-off addition of a new notification here.

So let's hold off on production implementation until we do our planning this coming week. We can consider this story as a spike.

**Comment by pdobacz** [comment]
:+1: copy that.

To wrap things up nicely, I've completed the upgrade of the PoC to "4+", which I already started experimenting with. Works fine but lots of tidies left behind. The code is in the same branch.

Dropping the TODO list here:

   - add a flag to the event denoting status
   - various refactoring FIXMEs throughout code
   - integration test the new events
   - unsubscription from the watcher to cause unsubscription from the notif. service. Requires managing channel's state
   - find for information/advice about how proper/hacky this PhoenixChannel-chaining is
   - update the NOTES and TODOS - especially ones about this being experimental etc etc
   - add proper configuration and usage docs (esp the `:notification_service_url` and `NOTIFICATION_SERVICE_URL` configs)

The effect is like this: (I only subscribe once to the `transfer` socket/channel **on the Watcher**). Note that there are two events, and the first one doesn't list the block information (block hash, submission height etc):

![Screenshot from 2019-04-26 11-30-30](https://user-images.githubusercontent.com/5735525/56798405-f32ea380-6816-11e9-9201-ad6ecbdd8fef.png)
